### ISSUE TITLE
Cross-architecture semantic inconsistency: attentions_output / mlps_output include the residual on BLOOM

### ISSUE THREAD

**Issue by hijohnnylin** [ndif-team/nnterp#51]
The following was a bug discovered by Claude while we were experimenting with different models. GPT 5.6 Sol has verified it as well

 From what I understand this does end up giiving values that many people may not intuitively expect (a judgment call whether or not this is truly "mislabeled" though), so it's up to you how much you want to weigh it. Thanks!

--

**Versions:** nnterp 1.3.0, nnsight 0.7.0, transformers 5.14.1, torch 2.13.0, CPU, float32

## Summary

On `bigscience/bloom-560m`, `model.attentions_output[layer]` and `model.mlps_output[layer]` expose residual-added states rather than additive attention and MLP contributions.

The accessors correctly return the primary tensor produced by the corresponding Hugging Face module. The problem is that this module boundary does not have consistent residual semantics across model families:

- On architectures where the residual addition happens in the transformer block, the attention or MLP module output is the additive sublayer contribution.
- On BLOOM, the residual addition happens inside `BloomAttention` and `BloomMLP`, so the module output is already an accumulated residual-stream state.

As a result, the same standardized accessor names silently denote mathematically different tensors across architectures while preserving the expected shape and dtype.

## Root cause

BLOOM performs the residual addition inside each sublayer:

```python
class BloomAttention(nn.Module):
    def forward(self, hidden_states, residual, ...):
        ...
        output_tensor = self.dense(context_layer)
        output_tensor = dropout_add(
            output_tensor,
            residual,
            self.hidden_dropout,
            self.training,
        )
        return output_tensor, ...

class BloomMLP(nn.Module):
    def forward(self, hidden_states, residual):
        ...
        intermediate_output = self.dense_4h_to_h(hidden_states)
        output = dropout_add(
            intermediate_output,
            residual,
            self.hidden_dropout,
            self.training,
        )
        return output
```

The general relationship is therefore:

```text
module output = residual argument + dropout(projected sublayer output)
```

For `bigscience/bloom-560m` in this reproduction:

- `apply_residual_connection_post_layernorm` is false, so the attention residual argument is `resid_pre` and the MLP residual argument is `resid_mid`.
- `hidden_dropout` is zero, so the output-projection tensor is also the additive contribution.

Consequently:

```text
BloomAttention output = resid_pre + attention contribution = resid_mid
BloomMLP output       = resid_mid + MLP contribution       = resid_post
```

A parameter named `residual` in a sublayer's `forward` signature is a useful architecture-discovery heuristic, but it should not be the semantic test by itself. The relevant condition is whether the returned tensor already includes a residual addition.

## Observed behavior

The standardized accessors are exactly equal to the residual-added Hugging Face module outputs and substantially different from the additive contributions.

Here, relative error is:

```text
||observed - reference||₂ / ||reference||₂
```

```text
layer 0:
  attentions_output vs attention contribution
    cos=+0.428220  rel=  2.0837  max_abs=8.486

  attentions_output vs BloomAttention module output
    cos=+1.000000  rel=  0.0000  max_abs=0

  mlps_output vs MLP contribution
    cos=+0.713971  rel=  1.1957  max_abs=8.604

  mlps_output vs BloomMLP module output
    cos=+1.000000  rel=  0.0000  max_abs=0


layer 12:
  attentions_output vs attention contribution
    cos=+0.054341  rel=131.2879  max_abs=516.9

  attentions_output vs BloomAttention module output
    cos=+1.000000  rel=  0.0000  max_abs=0

  mlps_output vs MLP contribution
    cos=+0.115973  rel= 50.4636  max_abs=517.0

  mlps_output vs BloomMLP module output
    cos=+1.000000  rel=  0.0000  max_abs=0


layer 23:
  attentions_output vs attention contribution
    cos=+0.353954  rel= 11.2409  max_abs=50.21

  attentions_output vs BloomAttention module output
    cos=+1.000000  rel=  0.0000  max_abs=0

  mlps_output vs MLP contribution
    cos=+0.140398  rel= 15.3787  max_abs=52.75

  mlps_output vs BloomMLP module output
    cos=+1.000000  rel=  0.0000  max_abs=0
```

This is not a numerical-precision difference. The accessor returns a different semantic tensor, and by the middle of the model the residual stream can dominate the additive contribution by two orders of magnitude.

## Minimal reproduction

This reproduction checks both attention and MLP at layer 12.

```python
import torch
from nnterp import StandardizedTransformer
from transformers import AutoModelForCausalLM, AutoTokenizer

MODEL = "bigscience/bloom-560m"
PROMPT = "The capital of France is Paris."
LAYER = 12


def unwrap(value):
    while isinstance(value, tuple):
        value = value[0]
    return value.detach().cpu()


# Raw Hugging Face reference.
tokenizer = AutoTokenizer.from_pretrained(MODEL)
hf_model = AutoModelForCausalLM.from_pretrained(
    MODEL,
    dtype=torch.float32,
).eval()

caught = {}
block = hf_model.transformer.h[LAYER]

# For this checkpoint hidden_dropout == 0, so these projection outputs
# are also the additive residual contributions.
block.self_attention.dense.register_forward_hook(
    lambda module, args, output: caught.__setitem__(
        "attention_contribution",
        output.detach().cpu(),
    )
)
block.self_attention.register_forward_hook(
    lambda module, args, output: caught.__setitem__(
        "attention_module_output",
        output[0].detach().cpu(),
    )
)
block.mlp.dense_4h_to_h.register_forward_hook(
    lambda module, args, output: caught.__setitem__(
        "mlp_contribution",
        output.detach().cpu(),
    )
)
block.mlp.register_forward_hook(
    lambda module, args, output: caught.__setitem__(
        "mlp_module_output",
        output.detach().cpu(),
    )
)

with torch.inference_mode():
    hf_model(**tokenizer(PROMPT, return_tensors="pt"))


# nnterp standardized accessors.
model = StandardizedTransformer(
    MODEL,
    dtype=torch.float32,
    device_map="cpu",
)

with model.trace(PROMPT):
    attention_output = model.attentions_output[LAYER].save()
    mlp_output = model.mlps_output[LAYER].save()

attention_output = unwrap(attention_output)
mlp_output = unwrap(mlp_output)

print(
    "attention accessor equals additive contribution:",
    torch.equal(attention_output, caught["attention_contribution"]),
)
print(
    "attention accessor equals literal module output:",
    torch.equal(attention_output, caught["attention_module_output"]),
)

print(
    "MLP accessor equals additive contribution:",
    torch.equal(mlp_output, caught["mlp_contribution"]),
)
print(
    "MLP accessor equals literal module output:",
    torch.equal(mlp_output, caught["mlp_module_output"]),
)
```

Observed output:

```text
attention accessor equals additive contribution: False
attention accessor equals literal module output: True
MLP accessor equals additive contribution: False
MLP accessor equals literal module output: True
```

## Why this matters

I understand that the current implementation exposes literal Hugging Face module I/O. The issue is that these standardized names can easily be interpreted and used as architecture-independent activation points.

For example, code may use `attentions_output` or `mlps_output` for:

- attribution over additive sublayer outputs;
- training an SAE on attention or MLP contributions;
- attention-output or MLP-output ablations;
- comparing the same conceptual activation across model families.

On BLOOM, such code instead receives `resid_mid` or `resid_post`. Nothing raises an error, and the shape, dtype, and sequence dimensions all appear correct.

## Suggested resolution

Possible resolutions include:

1. Add architecture-aware accessors such as:

   ```text
   attentions_contribution
   mlps_contribution
   ```

   These would consistently expose the tensors added to the residual stream.

2. Document explicitly that `attentions_output` and `mlps_output` mean literal Hugging Face module outputs and may therefore include residual additions, with warnings or architecture metadata for affected families.

3. If the intended contract is architecture-independent contribution semantics, map the accessors to a post-projection, post-dropout, pre-residual-addition boundary.

For BLOOM specifically, directly hooking `self_attention.dense` and `mlp.dense_4h_to_h` is sufficient for this checkpoint because hidden dropout is zero. A general implementation should account for dropout and should target the tensor immediately before residual addition rather than relying only on those projection-module names.

For read-only activation capture, the contributions could also be derived from adjacent residual states:

```python
attention_contribution = resid_mid - resid_pre
mlp_contribution = resid_post - resid_mid
```

For intervention APIs, however, a modified contribution would need to be inserted back into the residual-addition computation rather than exposed only as a derived value.

**Comment by Butanium** [comment]
@hijohnnylin thanks for reporting this! Does Claude think this could be detected at runtime by comparing the cosim sim between the layer and attention/MLP output?
